File: ztm_master_coding_interview_course.py/stacks_queues/stack_implementation.py
""" LIFO - Last In First Out """
from typing import Any


class Node:
    """ A class representing a Node since we are going to use linked list. """
    def __init__(self, data: Any) -> None:
        self.data = data
        self.next = None


class Stack:
    """ A class representing Stack data structure. """
    def __init__(self) -> None:
        self.top = None
        self.bottom = None
        self.length = 0

    
    def peek(self) -> Node | Exception:
        """ See which node is at the top position. """
        if self.top is None and self.bottom:
            self.bottom = None
            raise Exception("Stack already empty :d")
        else:
            return self.top.data
    

    def push(self, data: Any) -> None:
        """ Add elements to top position. """
        NEW_NODE = Node(data)

        if self.top is None:
            self.top = NEW_NODE
            self.bottom = NEW_NODE
        else:
            NEW_NODE.next = self.top
            self.bottom = self.top.next
            self.top = NEW_NODE
        
        self.length += 1


    def pop(self) -> None:
        """ Remove the last inserted Node which should be top. """
        self.top = self.top.next
        self.length -= 1


    def isempty(self) -> bool:
        """ Check if stack is empty or not. """
        if self.length > 0:
            return False
        else:
            return True


if __name__ == '__main__':
    stack_instance = Stack() # Instance stack class.

    # Add elements.
    stack_instance.push(1)
    stack_instance.push(2)
    stack_instance.push(3)

    # Remove elements and peek stack.
    stack_instance.pop()
    print(f"Current top -> {stack_instance.peek()}, bottom -> {stack_instance.bottom.data}\n")
    stack_instance.pop()

    # Remove last element, check if method is working, and force exception get raised.
    stack_instance.pop()
    print(f"Stack length -> {stack_instance.length}", stack_instance.isempty())
    print(stack_instance.peek())

File: ztm_master_coding_interview_course.py/stacks_queues/stack_implementation.js
/* LIFO - Last In First out. */

class Node {
    /* A class representing nodes of a linked list. */
    constructor (data) {
        this.data = data;
        this.next = null;
    }
}


class Stack {
    /* A class representing a stack data structure. */
    constructor () {
        this.top = null;
        this.bottom = null;
        this.length = 0;
    }

    peek () {
        /* Retrieve which node is at top position. */
        try {
            return this.top.data;
        } catch (error) {
            throw new Error("Stack already empty...");
        }
    }


    push (data) {
        /* Add nodes at top position. */
        const NEW_NODE = new Node(data);

        if (this.top === null) {
            this.top = NEW_NODE;
            this.bottom = NEW_NODE;
        } else {
            NEW_NODE.next = this.top;
            this.bottom = this.top.next;
            this.top = NEW_NODE;
        }
        this.length++;
    }


    pop () {
        /* Remove last element inserted. */
        this.top = this.top.next;
        this.length--;
    }


    isEmpty () {
        /* Check if stack is empty or not. */
        if (this.length > 0) {
            return false;
        } else {
            return true;
        }
    }

}


/* Example usage. */
const stackWithNodes = new Stack();

// Add elements.
stackWithNodes.push(1);
stackWithNodes.push(2);
stackWithNodes.push(3);

// Remove elements.
stackWithNodes.pop();
stackWithNodes.pop();

// Print out element at top position, and check if stack is empty...
console.log(stackWithNodes.peek(), stackWithNodes.isEmpty());

